controller/server_controller.py

```diff
@@ -125,6 +127,88 @@ def create_fn(labels, logger, name, namespace, spec, uid, **_):
     return {"createdResources": children_uids, "fullServerURL": get_urls(spec)[1]}
 
 
+def cull_idle_jupyter_servers(body, name, namespace, logger, **kwargs):
+    """
+    Check if a session is idle (has zero open connections in proxy and CPU is below
+    threshold). If the session is idle then annotate the jupyterserver resource with
+    the idle duration. If any sessions have been idle for long enough, then cull them.
+    """
+    try:
+        pod_name = body["status"]["mainPod"]["name"]
+    except KeyError:
+        return
+    cpu_usage = get_cpu_usage(pod=pod_name, namespace=namespace)
+    js_server_status = get_js_server_status(body)
+    custom_resource_api = get_api(config.api_version, config.custom_resource_name)
+    idle_seconds = int(body["status"].get("idleSeconds", 0))
+    logger.info(
+        f"Checking idle status of session {name}, "
+        f"idle seconds: {idle_seconds}, "
+        f"cpu usage: {cpu_usage}m, "
+        f"server status: {js_server_status}"
+    )
+
+    now = pytz.UTC.localize(datetime.utcnow())
+    jupyter_server_is_idle = (
+        cpu_usage <= config.CPU_USAGE_MILLICORES_IDLE_THRESHOLD
+        and type(js_server_status) is dict
+        and js_server_status.get("connections", 0) == 0
+        and js_server_status.get("kernels", 0) == 0
+        and (
+            now - js_server_status.get("last_activity", now).astimezone(pytz.UTC)
+        ).total_seconds()
+        > config.JUPYTER_SERVER_IDLE_CHECK_INTERVAL_SECONDS
+    )
+
+    if (
+        jupyter_server_is_idle
+        and idle_seconds + config.JUPYTER_SERVER_IDLE_CHECK_INTERVAL_SECONDS
+        >= config.JUPYTER_SERVER_CULL_IDLE_THRESHOLD_SECONDS
+    ):
+        logger.info(f"Deleting session {name} due to inactivity")
+        try:
+            custom_resource_api.delete(name=name, namespace=namespace)
+        except NotFoundError:
```

Comment:
I think this is a bit generous and has the potential of hiding actual bugs. Catching the `NotFoundError` makes sense to me, but I'd at least log a warning along the lines of: "Trying to delete jupyter server X in namespace Y, but we can not find it. Has it been deleted in the mean time?")

Reply:
Same applies for the patch requests further down.